### ISSUE TITLE
RecursionError in plain python turns into segmentation fault in terminal ipython

### ISSUE THREAD

**Issue by aarchiba** [ipython/ipython#12846]
The following program causes a RecursionError due to a numpy bug(?) https://github.com/numpy/numpy/issues/18548 :
```python
import numpy as np
None + np.longdouble(0)
```
In plain python:
```
$ python ../zot.py 
Traceback (most recent call last):
  File "../zot.py", line 2, in <module>
    None + np.longdouble(0)
RecursionError: maximum recursion depth exceeded while calling a Python object
```
In terminal ipython:
```
$ ipython
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.17.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %run ../zot.py
Segmentation fault (core dumped)
```
Also if it helps
```
n [3]: numpy.version.version
Out[3]: '1.19.1'
```
The machine is
```
$ uname -a
Linux dop307 5.4.0-65-generic #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by seberg** [comment]
Just to note, I doubt that IPython can do anything about the recursion error going to a segfault.  There was a PEP just shot down that tried to do something about it (as far as I understood it), but I guess it was too much magic, complexity, and I am not certain it would solve this particular path anyway.

Windows users might even get a semi-reasonable error here.  The only difference between Python and IPython here is that IPython sets `sys.setrecursionlimit(3000)` and Python defaults to 1000. The recursion here goes into the NumPy ufunc machinery, which is probably fairly stack-heavy, so the segfault is probably just running out of stack-space here :(.

**Comment by MrMino** [comment]
TBH it feels like a bug in numpy.

```
>>> from numpy import longdouble
>>> longdouble(0) + None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'numpy.float128' and 'NoneType'
>>> None + longdouble(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RecursionError: maximum recursion depth exceeded while calling a Python object
```

`longdouble.__add__` properly raises where it is incompatible with its args, while `longdouble.__radd__` recurses infinitely. That shouldn't happen.

**Comment by MrMino** [comment]
TBH to me it looks like a bug in numpy.

```
>>> from numpy import longdouble
>>> longdouble(0) + None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'numpy.float128' and 'NoneType'
>>> None + longdouble(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RecursionError: maximum recursion depth exceeded while calling a Python object
```

`longdouble.__add__` properly raises where it is incompatible with its args, while `longdouble.__radd__` recurses infinitely. That shouldn't happen.

**Comment by MrMino** [comment]
I'd say that this issue is a no-op for IPython, given the fix was already made in numpy/numpy#18548 (nice BTW, @aarchiba).

Actually, Python interpreter does the same thing, if reproduced with higher recursion limit:

```
>>> import sys
>>> sys.setrecursionlimit(3000)
>>> import numpy as np 
>>> None + np.longdouble(0)
[1]    2137395 segmentation fault (core dumped)  python
```

Unless I'm missing something, it is not possible for the Python code to shield against whatever mischief the C code is up to at any given moment.
